src/libraries/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -136,42 +140,84 @@ public unsafe BitArray(bool[] values)
+                        Vector256<byte> isFalse = Avx2.CompareEqual(vector, zero);
                         int result = Avx2.MoveMask(isFalse);
                         m_array[i / 32] = ~result;
                     }
                 }
             }
             else if (Sse2.IsSupported)
             {
+                // JIT does not support code hoisting for SIMD yet
+                Vector128<byte> zero = Vector128<byte>.Zero;
                 fixed (bool* ptr = values)
                 {
-                    for (; (i + Vector128<byte>.Count * 2) <= values.Length; i += Vector128<byte>.Count * 2)
+                    for (; (i + Vector128<byte>.Count * 2) <= values.Length; i += (uint)Vector128<byte>.Count * 2)
                     {
                         Vector128<byte> lowerVector = Sse2.LoadVector128((byte*)ptr + i);
-                        Vector128<byte> lowerIsFalse = Sse2.CompareEqual(lowerVector, Vector128<byte>.Zero);
+                        Vector128<byte> lowerIsFalse = Sse2.CompareEqual(lowerVector, zero);
                         int lowerPackedIsFalse = Sse2.MoveMask(lowerIsFalse);
 
                         Vector128<byte> upperVector = Sse2.LoadVector128((byte*)ptr + i + Vector128<byte>.Count);
-                        Vector128<byte> upperIsFalse = Sse2.CompareEqual(upperVector, Vector128<byte>.Zero);
+                        Vector128<byte> upperIsFalse = Sse2.CompareEqual(upperVector, zero);
                         int upperPackedIsFalse = Sse2.MoveMask(upperIsFalse);
 
                         m_array[i / 32] = ~((upperPackedIsFalse << 16) | lowerPackedIsFalse);
                     }
                 }
             }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                // JIT does not support code hoisting for SIMD yet
+                // However comparison against zero can be replaced to cmeq against zero (vceqzq_s8)
+                // See dotnet/runtime#33972 for details
+                Vector128<byte> zero = Vector128<byte>.Zero;
+                fixed (bool* ptr = values)
+                {
+                    for (; (i + Vector128<byte>.Count * 2) <= values.Length; i += (uint)Vector128<byte>.Count * 2)
+                    {
+                        // Same logic as SSE2 path, however we lack MoveMask (equivalent) instruction
+                        // As a workaround, mask out the relevant bit after comparison
+                        // and combine by ORing all of them together (In this case, adding all of them does the same thing)
+                        Vector128<byte> lowerVector = AdvSimd.LoadVector128((byte*)ptr + i);
+                        Vector128<byte> lowerIsFalse = AdvSimd.CompareEqual(lowerVector, zero);
+                        Vector128<byte> bitsExtracted1 = AdvSimd.And(lowerIsFalse, s_bitMask128);
+                        bitsExtracted1 = AdvSimd.Arm64.AddPairwise(bitsExtracted1, bitsExtracted1);
```

Comment:
Is 3 add pairwise more efficient than an add across and duplicate?

Reply:
> Is 3 add pairwise more efficient than an add across and duplicate?

That answer depends on which micro-architecture. On older cores like the A57 yes, on pure cycles the `addp` wins hands down as `addv` on18B is very very expensive. and then you have the `dup` on top of it which is also expensive..

On more modern micro-architectures like `Neoverse-N1` the `addv + dup` sequence juuust about wins out. If you're targeting the widest range of cores then I'd go with the 3 `addp`.

Reply:
Thanks for the explanation! It might be worth having a small comment about that here.

I think we are using the CORTEX-A55 as the baseline for our "performance" cycle metrics.

Reply:
Ah, `Cortex-A55` is an interesting one, since it's new enough that the performance with `addv` has been improved a lot, but `addp` wasn't improved yet. On an `A55` the `addv+dup` would win out with ~1/3rd margin. So in this case it's a bit of a difficult question, without a cost model one core will always lose :(